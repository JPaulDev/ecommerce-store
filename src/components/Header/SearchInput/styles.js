import styled from 'styled-components';

export const Form = styled.form`
  justify-self: center;
  position: relative;
  width: 100%;

  &::after {
    border-radius: 9px;
    box-shadow: 0 0 0 3px var(--color-orange-1);
    content: '';
    height: 100%;
    left: 0;
    opacity: ${({ isFocused }) => (isFocused ? '1' : '0')};
    pointer-events: none;
    position: absolute;
    transition: opacity 150ms ease-in-out;
    width: 100%;
  }

  @media screen and (min-width: 545px) {
    max-width: 420px;
  }

  @media screen and (max-width: 929px) {
    grid-column: 1/3;
    grid-row: 2/3;
    margin-top: 15px;
  }
`;

export const Input = styled.input`
  border-radius: 8px;
  color: var(--color-grey-2);
  font-size: 0.94rem;
  font-weight: 500;
  min-height: 44px;
  outline: none;
  padding-left: 10px;
  padding-right: 65px;
  width: 100%;

  &::placeholder {
    font-size: 0.94rem;
  }
`;

export const Button = styled.button`
  align-items: center;
  background-color: var(--color-orange-1);
  border-radius: 8px;
  cursor: pointer;
  display: flex;
  height: 100%;
  justify-content: center;
  padding-left: 3px;
  position: absolute;
  right: 0;
  top: 0;
  width: 55px;
`;
