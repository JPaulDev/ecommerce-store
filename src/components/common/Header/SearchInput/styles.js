import styled, { css } from 'styled-components';

export const Form = styled.form`
  justify-self: center;
  position: relative;
  width: 100%;

  &::after {
    border-radius: var(--border-radius-xl);
    box-shadow: 0 0 0 3px var(--color-orange-1);
    content: '';
    height: 100%;
    left: 0;
    opacity: ${({ isFocused }) => (isFocused ? '1' : '0')};
    pointer-events: none;
    position: absolute;
    transition: opacity 150ms ease-in-out;
    width: 100%;
  }

  @media screen and (min-width: 545px) {
    max-width: 420px;
  }

  @media screen and (max-width: 934px) {
    grid-column: 1 / 3;
    grid-row: 2 / 3;
    margin-top: 15px;
  }
`;

export const Input = css`
  border-radius: var(--border-radius-xl);
  color: var(--color-grey-2);
  font-size: var(--font-size-14);
  font-weight: var(--font-weight-medium);
  min-height: 44px;
  outline: none;
  padding-left: 10px;
  padding-right: 65px;
  width: 100%;

  &::-webkit-search-cancel-button {
    display: none;
  }
`;

export const Button = styled.button`
  align-items: center;
  background-color: var(--color-orange-1);
  border-radius: var(--border-radius-xl);
  cursor: pointer;
  display: flex;
  height: 100%;
  justify-content: center;
  padding-left: 3px;
  position: absolute;
  right: 0;
  top: 0;
  width: 55px;
`;
