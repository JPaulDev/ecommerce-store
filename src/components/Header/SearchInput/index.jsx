import { useState, useRef, useId, useEffect } from 'react';
import { ReactComponent as SearchIcon } from '../../../assets/svg/search.svg';
import * as Styled from './styles';

export default function SearchInput() {
  const [query, setQuery] = useState('');
  const [isFocused, setIsFocused] = useState(false);
  const ref = useRef(null);
  const id = useId();

  const handleChange = (e) => setQuery(e.target.value);

  useEffect(() => {
    const element = ref.current;
    const handleFocus = () => setIsFocused(true);
    const handleBlur = () => setIsFocused(false);

    element.addEventListener('focus', handleFocus);
    element.addEventListener('blur', handleBlur);

    return () => {
      element.removeEventListener('focus', handleFocus);
      element.removeEventListener('blur', handleBlur);
    };
  }, []);

  return (
    <Styled.Form
      role="search"
      action="/search"
      method="get"
      isFocused={isFocused}
    >
      <label htmlFor={id} hidden>
        Search
      </label>
      <Styled.Input
        ref={ref}
        id={id}
        value={query}
        name="searchQuery"
        placeholder="Search PC-Connect.co.uk..."
        type="search"
        autoComplete="off"
        maxLength="50"
        required
        onChange={handleChange}
      />
      <Styled.Button title="Search" type="submit">
        <SearchIcon width={30} height={30} fill="white" />
      </Styled.Button>
    </Styled.Form>
  );
}
