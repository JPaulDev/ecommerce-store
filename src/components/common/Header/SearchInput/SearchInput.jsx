import { useState } from 'react';
import { InputWithLabel } from '../../../ui';
import { Search } from '../../../icons';
import * as Styled from './styles';

export default function SearchInput() {
  const [query, setQuery] = useState('');
  const [isFocused, setIsFocused] = useState(false);

  const handleChangeSearchQuery = (e) => setQuery(e.target.value);

  const handleChangeFocus = () => setIsFocused(!isFocused);

  return (
    <Styled.Form
      role="search"
      action="/search"
      method="get"
      isFocused={isFocused}
    >
      <InputWithLabel
        label="Search:"
        styles={Styled.Input}
        name="searchQuery"
        type="search"
        placeholder="Search PC-Connect.co.uk..."
        maxLength="50"
        required
        aria-required
        onChange={handleChangeSearchQuery}
        onFocus={handleChangeFocus}
        onBlur={handleChangeFocus}
        value={query}
      />
      <Styled.Button title="Search" aria-label="Search" type="submit">
        <Search width={30} height={30} />
      </Styled.Button>
    </Styled.Form>
  );
}
