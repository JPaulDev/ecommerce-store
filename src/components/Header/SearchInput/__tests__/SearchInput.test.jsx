import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import SearchInput from '../index';

it('search input correctly registers user input', () => {
  render(<SearchInput />);
  const input = screen.getByRole('searchbox');
  userEvent.type(input, 'processors');

  expect(input.value).toBe('processors');
});

it('search input renders correctly', () => {
  const tree = renderer.create(<SearchInput />).toJSON();
  expect(tree).toMatchSnapshot();
});
