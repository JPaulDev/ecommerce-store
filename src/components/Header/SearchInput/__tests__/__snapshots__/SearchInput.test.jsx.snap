// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component renders a search input with a button and a search icon 1`] = `
<form
  action="/search"
  className="sc-bczRLJ bDMbYN"
  method="get"
  role="search"
>
  <label
    hidden={true}
    htmlFor=":r0:"
  >
    Search
  </label>
  <input
    autoComplete="off"
    className="sc-gsnTZi iltnbf"
    id=":r0:"
    maxLength="50"
    name="searchQuery"
    onChange={[Function]}
    placeholder="Search PC-Connect.co.uk..."
    required={true}
    type="search"
    value=""
  />
  <button
    className="sc-dkzDqf gfrkhO"
    title="Search"
    type="submit"
  >
    <svg
      fill="white"
      height="30px"
      width="30px"
    >
      search.svg
    </svg>
  </button>
</form>
`;
